high cpu usage when using `- MAX_RETRIES=0`

i have notice when i use `- MAX_RETRIES=0` there is an unusual cpu usage, any thoughts?

@sonroyaalmerol trying ur new image at the moment. seem to be working although i did catch this in the logs
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5fe772]
goroutine 54271 [running]:
bufio.(*Writer).Flush(0x466545?)
	/usr/local/go/src/bufio/bufio.go:633 +0x12
net/http.(*chunkWriter).flush(0x40b2ff?)
	/usr/local/go/src/net/http/server.go:403 +0x39
net/http.(*response).FlushError(0xc0004600e0)
	/usr/local/go/src/net/http/server.go:1773 +0x51
net/http.(*response).Flush(0xaeed10?)
	/usr/local/go/src/net/http/server.go:1765 +0x13
m3u-stream-merger/proxy.(*StreamInstance).ProxyStream.func2()
	/app/proxy/proxy_stream.go:87 +0x63
m3u-stream-merger/proxy.(*StreamInstance).ProxyStream(0xc005c620c0, {0x8abc20, 0xc000410cd0}, 0x0, 0xc0000c6000, 0xc001b79040, {0x8ab0a8, 0xc0004600e0}, 0xc000ae90a0)
	/app/proxy/proxy_stream.go:143 +0xb2a
created by m3u-stream-merger/handlers.StreamHandler in goroutine 3512
	/app/handlers/stream_handler.go:80 +0x6fe
```

--- a/handlers/stream_handler.go
+++ b/handlers/stream_handler.go
@@ -47,7 +47,7 @@ func StreamHandler(w http.ResponseWriter, r *http.Request, cm *store.Concurrency
 			utils.SafeLogf("Client disconnected: %s\n", r.RemoteAddr)
 			return
 		default:
-			resp, selectedUrl, selectedIndex, err = stream.LoadBalancer(&testedIndexes, r.Method)
+			resp, selectedUrl, selectedIndex, err = stream.LoadBalancer(ctx, &testedIndexes, r.Method)
 			if err != nil {
 				utils.SafeLogf("Error reloading stream for %s: %v\n", streamUrl, err)
 				return
@@ -78,25 +78,32 @@ func StreamHandler(w http.ResponseWriter, r *http.Request, cm *store.Concurrency
 			go stream.ProxyStream(ctx, selectedIndex, resp, r, w, exitStatus)
 			testedIndexes = append(testedIndexes, selectedIndex)
 
-			streamExitCode := <-exitStatus
-			utils.SafeLogf("Exit code %d received from %s\n", streamExitCode, selectedUrl)
-
-			if streamExitCode == 2 && utils.EOFIsExpected(resp) {
-				utils.SafeLogf("Successfully proxied playlist: %s\n", r.RemoteAddr)
-				cancel()
-			} else if streamExitCode == 1 || streamExitCode == 2 {
-				// Retry on server-side connection errors
-				utils.SafeLogf("Retrying other servers...\n")
-			} else if streamExitCode == 4 {
-				utils.SafeLogf("Finished handling %s request: %s\n", r.Method, r.RemoteAddr)
-				cancel()
-			} else {
-				// Consider client-side connection errors as complete closure
+			select {
+			case <-ctx.Done():
 				utils.SafeLogf("Client has closed the stream: %s\n", r.RemoteAddr)
-				cancel()
-			}
+				resp.Body.Close()
 
-			resp.Body.Close()
+				return
+			case streamExitCode := <-exitStatus:
+				utils.SafeLogf("Exit code %d received from %s\n", streamExitCode, selectedUrl)
+
+				if streamExitCode == 2 && utils.EOFIsExpected(resp) {
+					utils.SafeLogf("Successfully proxied playlist: %s\n", r.RemoteAddr)
+					cancel()
+				} else if streamExitCode == 1 || streamExitCode == 2 {
+					// Retry on server-side connection errors
+					utils.SafeLogf("Retrying other servers...\n")
+				} else if streamExitCode == 4 {
+					utils.SafeLogf("Finished handling %s request: %s\n", r.Method, r.RemoteAddr)
+					cancel()
+				} else {
+					// Consider client-side connection errors as complete closure
+					utils.SafeLogf("Unable to write to client. Assuming stream has been closed: %s\n", r.RemoteAddr)
+					cancel()
+				}
+
+				resp.Body.Close()
+			}
 		}
 	}
 }

--- a/handlers/stream_handler.go
+++ b/handlers/stream_handler.go
@@ -1,7 +1,6 @@
 package handlers
 
 import (
-	"context"
 	"m3u-stream-merger/proxy"
 	"m3u-stream-merger/store"
 	"m3u-stream-merger/utils"
@@ -14,8 +13,7 @@ import (
 func StreamHandler(w http.ResponseWriter, r *http.Request, cm *store.ConcurrencyManager) {
 	debug := os.Getenv("DEBUG") == "true"
 
-	ctx, cancel := context.WithCancel(r.Context())
-	defer cancel()
+	ctx := r.Context()
 
 	utils.SafeLogf("Received request from %s for URL: %s\n", r.RemoteAddr, r.URL.Path)
 
@@ -89,17 +87,17 @@ func StreamHandler(w http.ResponseWriter, r *http.Request, cm *store.Concurrency
 
 				if streamExitCode == 2 && utils.EOFIsExpected(resp) {
 					utils.SafeLogf("Successfully proxied playlist: %s\n", r.RemoteAddr)
-					cancel()
+					return
 				} else if streamExitCode == 1 || streamExitCode == 2 {
 					// Retry on server-side connection errors
 					utils.SafeLogf("Retrying other servers...\n")
 				} else if streamExitCode == 4 {
 					utils.SafeLogf("Finished handling %s request: %s\n", r.Method, r.RemoteAddr)
-					cancel()
+					return
 				} else {
 					// Consider client-side connection errors as complete closure
 					utils.SafeLogf("Unable to write to client. Assuming stream has been closed: %s\n", r.RemoteAddr)
-					cancel()
+					return
 				}
 
 				resp.Body.Close()

--- a/handlers/stream_handler.go
+++ b/handlers/stream_handler.go
@@ -38,6 +38,11 @@ func StreamHandler(w http.ResponseWriter, r *http.Request, cm *store.Concurrency
 	firstWrite := true
 
 	var resp *http.Response
+	defer func() {
+		if resp != nil && resp.Body != nil {
+			resp.Body.Close()
+		}
+	}()
 
 	for {
 		select {
@@ -45,6 +50,10 @@ func StreamHandler(w http.ResponseWriter, r *http.Request, cm *store.Concurrency
 			utils.SafeLogf("Client disconnected: %s\n", r.RemoteAddr)
 			return
 		default:
+			if resp != nil && resp.Body != nil {
+				resp.Body.Close()
+			}
+
 			resp, selectedUrl, selectedIndex, err = stream.LoadBalancer(ctx, &testedIndexes, r.Method)
 			if err != nil {
 				utils.SafeLogf("Error reloading stream for %s: %v\n", streamUrl, err)
@@ -79,8 +88,6 @@ func StreamHandler(w http.ResponseWriter, r *http.Request, cm *store.Concurrency
 			select {
 			case <-ctx.Done():
 				utils.SafeLogf("Client has closed the stream: %s\n", r.RemoteAddr)
-				resp.Body.Close()
-
 				return
 			case streamExitCode := <-exitStatus:
 				utils.SafeLogf("Exit code %d received from %s\n", streamExitCode, selectedUrl)
@@ -99,8 +106,6 @@ func StreamHandler(w http.ResponseWriter, r *http.Request, cm *store.Concurrency
 					utils.SafeLogf("Unable to write to client. Assuming stream has been closed: %s\n", r.RemoteAddr)
 					return
 				}
-
-				resp.Body.Close()
 			}
 		}
 	}

--- a/handlers/stream_handler.go
+++ b/handlers/stream_handler.go
@@ -73,7 +73,6 @@ func StreamHandler(w http.ResponseWriter, r *http.Request, cm *store.Concurrency
 			}
 
 			exitStatus := make(chan int)
-			defer close(exitStatus)
 
 			utils.SafeLogf("Proxying %s to %s\n", r.RemoteAddr, selectedUrl)
 			go stream.ProxyStream(ctx, selectedIndex, resp, r, w, exitStatus)

--- a/handlers/stream_handler.go
+++ b/handlers/stream_handler.go
@@ -73,6 +73,7 @@ func StreamHandler(w http.ResponseWriter, r *http.Request, cm *store.Concurrency
 			}
 
 			exitStatus := make(chan int)
+			defer close(exitStatus)
 
 			utils.SafeLogf("Proxying %s to %s\n", r.RemoteAddr, selectedUrl)
 			go stream.ProxyStream(ctx, selectedIndex, resp, r, w, exitStatus)

--- a/handlers/stream_handler.go
+++ b/handlers/stream_handler.go
@@ -45,67 +45,61 @@ func StreamHandler(w http.ResponseWriter, r *http.Request, cm *store.Concurrency
 	}()
 
 	for {
+		if resp != nil && resp.Body != nil {
+			resp.Body.Close()
+		}
+
+		resp, selectedUrl, selectedIndex, err = stream.LoadBalancer(ctx, &testedIndexes, r.Method)
+		if err != nil {
+			utils.SafeLogf("Error reloading stream for %s: %v\n", streamUrl, err)
+			return
+		}
+
+		// HTTP header initialization
+		if firstWrite {
+			for k, v := range resp.Header {
+				if strings.ToLower(k) == "content-length" {
+					continue
+				}
+
+				for _, val := range v {
+					w.Header().Set(k, val)
+				}
+			}
+			w.WriteHeader(resp.StatusCode)
+
+			if debug {
+				utils.SafeLogf("[DEBUG] Headers set for response: %v\n", w.Header())
+			}
+			firstWrite = false
+		}
+
+		exitStatus := make(chan int)
+
+		utils.SafeLogf("Proxying %s to %s\n", r.RemoteAddr, selectedUrl)
+		go stream.ProxyStream(ctx, selectedIndex, resp, r, w, exitStatus)
+		testedIndexes = append(testedIndexes, selectedIndex)
+
 		select {
 		case <-ctx.Done():
-			utils.SafeLogf("Client disconnected: %s\n", r.RemoteAddr)
+			utils.SafeLogf("Client has closed the stream: %s\n", r.RemoteAddr)
 			return
-		default:
-			if resp != nil && resp.Body != nil {
-				resp.Body.Close()
-			}
+		case streamExitCode := <-exitStatus:
+			utils.SafeLogf("Exit code %d received from %s\n", streamExitCode, selectedUrl)
 
-			resp, selectedUrl, selectedIndex, err = stream.LoadBalancer(ctx, &testedIndexes, r.Method)
-			if err != nil {
-				utils.SafeLogf("Error reloading stream for %s: %v\n", streamUrl, err)
+			if streamExitCode == 2 && utils.EOFIsExpected(resp) {
+				utils.SafeLogf("Successfully proxied playlist: %s\n", r.RemoteAddr)
 				return
-			}
-
-			// HTTP header initialization
-			if firstWrite {
-				for k, v := range resp.Header {
-					if strings.ToLower(k) == "content-length" {
-						continue
-					}
-
-					for _, val := range v {
-						w.Header().Set(k, val)
-					}
-				}
-				w.WriteHeader(resp.StatusCode)
-
-				if debug {
-					utils.SafeLogf("[DEBUG] Headers set for response: %v\n", w.Header())
-				}
-				firstWrite = false
-			}
-
-			exitStatus := make(chan int)
-
-			utils.SafeLogf("Proxying %s to %s\n", r.RemoteAddr, selectedUrl)
-			go stream.ProxyStream(ctx, selectedIndex, resp, r, w, exitStatus)
-			testedIndexes = append(testedIndexes, selectedIndex)
-
-			select {
-			case <-ctx.Done():
-				utils.SafeLogf("Client has closed the stream: %s\n", r.RemoteAddr)
+			} else if streamExitCode == 1 || streamExitCode == 2 {
+				// Retry on server-side connection errors
+				utils.SafeLogf("Retrying other servers...\n")
+			} else if streamExitCode == 4 {
+				utils.SafeLogf("Finished handling %s request: %s\n", r.Method, r.RemoteAddr)
 				return
-			case streamExitCode := <-exitStatus:
-				utils.SafeLogf("Exit code %d received from %s\n", streamExitCode, selectedUrl)
-
-				if streamExitCode == 2 && utils.EOFIsExpected(resp) {
-					utils.SafeLogf("Successfully proxied playlist: %s\n", r.RemoteAddr)
-					return
-				} else if streamExitCode == 1 || streamExitCode == 2 {
-					// Retry on server-side connection errors
-					utils.SafeLogf("Retrying other servers...\n")
-				} else if streamExitCode == 4 {
-					utils.SafeLogf("Finished handling %s request: %s\n", r.Method, r.RemoteAddr)
-					return
-				} else {
-					// Consider client-side connection errors as complete closure
-					utils.SafeLogf("Unable to write to client. Assuming stream has been closed: %s\n", r.RemoteAddr)
-					return
-				}
+			} else {
+				// Consider client-side connection errors as complete closure
+				utils.SafeLogf("Unable to write to client. Assuming stream has been closed: %s\n", r.RemoteAddr)
+				return
 			}
 		}
 	}

--- a/proxy/load_balancer.go
+++ b/proxy/load_balancer.go
@@ -11,6 +11,7 @@ import (
 	"sort"
 	"strconv"
 	"strings"
+	"time"
 )
 
 type StreamInstance struct {
@@ -46,6 +47,11 @@ func (instance *StreamInstance) LoadBalancer(ctx context.Context, previous *[]in
 	}
 
 	lap := 0
+
+	// Backoff settings
+	initialBackoff := 200 * time.Millisecond
+	maxBackoff := 2 * time.Second
+	currentBackoff := initialBackoff
 
 	for lap < maxLaps || maxLaps == 0 {
 		if debug {
@@ -96,6 +102,15 @@ func (instance *StreamInstance) LoadBalancer(ctx context.Context, previous *[]in
 				*previous = []int{}
 			}
 
+		}
+
+		select {
+		case <-time.After(currentBackoff):
+			currentBackoff *= 2
+			if currentBackoff > maxBackoff {
+				currentBackoff = maxBackoff
+			}
+		case <-ctx.Done():
 		}
 
 		lap++

--- a/proxy/load_balancer.go
+++ b/proxy/load_balancer.go
@@ -1,6 +1,7 @@
 package proxy
 
 import (
+	"context"
 	"fmt"
 	"m3u-stream-merger/store"
 	"m3u-stream-merger/utils"
@@ -29,7 +30,7 @@ func NewStreamInstance(streamUrl string, cm *store.ConcurrencyManager) (*StreamI
 	}, nil
 }
 
-func (instance *StreamInstance) LoadBalancer(previous *[]int, method string) (*http.Response, string, int, error) {
+func (instance *StreamInstance) LoadBalancer(ctx context.Context, previous *[]int, method string) (*http.Response, string, int, error) {
 	debug := os.Getenv("DEBUG") == "true"
 
 	m3uIndexes := utils.GetM3UIndexes()
@@ -52,43 +53,49 @@ func (instance *StreamInstance) LoadBalancer(previous *[]int, method string) (*h
 		}
 		allSkipped := true // Assume all URLs might be skipped
 
-		for _, index := range m3uIndexes {
-			if slices.Contains(*previous, index) {
-				utils.SafeLogf("Skipping M3U_%d: marked as previous stream\n", index+1)
-				continue
-			}
-
-			url, ok := instance.Info.URLs[index]
-			if !ok {
-				utils.SafeLogf("Channel not found from M3U_%d: %s\n", index+1, instance.Info.Title)
-				continue
-			}
-
-			if instance.Cm.CheckConcurrency(index) {
-				utils.SafeLogf("Concurrency limit reached for M3U_%d: %s\n", index+1, url)
-				continue
-			}
-
-			allSkipped = false // At least one URL is not skipped
-
-			resp, err := utils.CustomHttpRequest(method, url)
-			if err == nil {
-				if debug {
-					utils.SafeLogf("[DEBUG] Successfully fetched stream from %s\n", url)
+		select {
+		case <-ctx.Done():
+			return nil, "", -1, fmt.Errorf("Cancelling load balancer.")
+		default:
+			for _, index := range m3uIndexes {
+				if slices.Contains(*previous, index) {
+					utils.SafeLogf("Skipping M3U_%d: marked as previous stream\n", index+1)
+					continue
 				}
-				return resp, url, index, nil
-			}
-			utils.SafeLogf("Error fetching stream: %s\n", err.Error())
-			if debug {
-				utils.SafeLogf("[DEBUG] Error fetching stream from %s: %s\n", url, err.Error())
-			}
-		}
 
-		if allSkipped {
-			if debug {
-				utils.SafeLogf("[DEBUG] All streams skipped in lap %d\n", lap)
+				url, ok := instance.Info.URLs[index]
+				if !ok {
+					utils.SafeLogf("Channel not found from M3U_%d: %s\n", index+1, instance.Info.Title)
+					continue
+				}
+
+				if instance.Cm.CheckConcurrency(index) {
+					utils.SafeLogf("Concurrency limit reached for M3U_%d: %s\n", index+1, url)
+					continue
+				}
+
+				allSkipped = false // At least one URL is not skipped
+
+				resp, err := utils.CustomHttpRequest(method, url)
+				if err == nil {
+					if debug {
+						utils.SafeLogf("[DEBUG] Successfully fetched stream from %s\n", url)
+					}
+					return resp, url, index, nil
+				}
+				utils.SafeLogf("Error fetching stream: %s\n", err.Error())
+				if debug {
+					utils.SafeLogf("[DEBUG] Error fetching stream from %s: %s\n", url, err.Error())
+				}
 			}
-			*previous = []int{}
+
+			if allSkipped {
+				if debug {
+					utils.SafeLogf("[DEBUG] All streams skipped in lap %d\n", lap)
+				}
+				*previous = []int{}
+			}
+
 		}
 
 		lap++

--- a/proxy/proxy_stream.go
+++ b/proxy/proxy_stream.go
@@ -83,9 +83,6 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 
 	defer func() {
 		buffer = nil
-		if flusher, ok := w.(http.Flusher); ok {
-			flusher.Flush()
-		}
 	}()
 
 	timeoutSecond := 3

--- a/proxy/proxy_stream.go
+++ b/proxy/proxy_stream.go
@@ -14,7 +14,6 @@ import (
 )
 
 func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, resp *http.Response, r *http.Request, w http.ResponseWriter, statusChan chan int) {
-	debug := os.Getenv("DEBUG") == "true"
 	bufferMbInt, err := strconv.Atoi(os.Getenv("BUFFER_MB"))
 	if err != nil || bufferMbInt < 0 {
 		bufferMbInt = 0
@@ -82,9 +81,9 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 		}
 	}()
 
-	timeoutSecond, err := strconv.Atoi(os.Getenv("STREAM_TIMEOUT"))
-	if err != nil || timeoutSecond < 0 {
-		timeoutSecond = 3
+	timeoutSecond := 3
+	if ts, err := strconv.Atoi(os.Getenv("STREAM_TIMEOUT")); err == nil && ts >= 0 {
+		timeoutSecond = ts
 	}
 
 	timeoutDuration := time.Duration(timeoutSecond) * time.Second
@@ -94,16 +93,14 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 	timer := time.NewTimer(timeoutDuration)
 	defer timer.Stop()
 
-	// Backoff settings
-	initialBackoff := 200 * time.Millisecond
-	maxBackoff := time.Duration(timeoutSecond-1) * time.Second
-	currentBackoff := initialBackoff
+	timeStarted := time.Now()
+	lastErr := timeStarted
 
 	returnStatus := 0
 
 	for {
 		select {
-		case <-ctx.Done(): // handle context cancellation
+		case <-ctx.Done():
 			utils.SafeLogf("Context canceled for stream: %s\n", r.RemoteAddr)
 			statusChan <- 0
 			return
@@ -113,52 +110,39 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 			return
 		default:
 			n, err := resp.Body.Read(buffer)
-			if err != nil {
-				if err == io.EOF {
-					utils.SafeLogf("Stream ended (EOF reached): %s\n", r.RemoteAddr)
-					if utils.EOFIsExpected(resp) || timeoutSecond == 0 {
-						statusChan <- 2
-						return
-					}
-
-					returnStatus = 2
-					utils.SafeLogf("Retrying same stream until timeout (%d seconds) is reached...\n", timeoutSecond)
-					if debug {
-						utils.SafeLogf("[DEBUG] Retrying same stream with backoff of %v...\n", currentBackoff)
-					}
-
-					time.Sleep(currentBackoff)
-					currentBackoff *= 2
-					if currentBackoff > maxBackoff {
-						currentBackoff = maxBackoff
-					}
-
-					continue
+			switch {
+			case err == io.EOF:
+				lastErr = time.Now()
+				if utils.EOFIsExpected(resp) || timeoutSecond == 0 {
+					utils.SafeLogf("Stream ended (expected EOF reached): %s\n", r.RemoteAddr)
+					statusChan <- 2
+					return
 				}
 
+				utils.SafeLogf("Stream ended (unexpected EOF reached): %s\n", r.RemoteAddr)
+				returnStatus = 2
+
+				utils.SafeLogf("Retrying same stream until timeout (%d seconds) is reached...\n", timeoutSecond)
+				time.Sleep(time.Millisecond * 500)
+
+				continue
+			case err != nil:
+				lastErr = time.Now()
 				utils.SafeLogf("Error reading stream: %s\n", err.Error())
-
 				returnStatus = 1
-
 				if timeoutSecond == 0 {
 					statusChan <- 1
 					return
 				}
 
-				if debug {
-					utils.SafeLogf("[DEBUG] Retrying same stream with backoff of %v...\n", currentBackoff)
-				}
-
-				time.Sleep(currentBackoff)
-				currentBackoff *= 2
-				if currentBackoff > maxBackoff {
-					currentBackoff = maxBackoff
-				}
+				utils.SafeLogf("Retrying same stream until timeout (%d seconds) is reached...\n", timeoutSecond)
+				time.Sleep(time.Millisecond * 500)
 
 				continue
 			}
 
 			if _, err := w.Write(buffer[:n]); err != nil {
+				lastErr = time.Now()
 				utils.SafeLogf("Error writing to response: %s\n", err.Error())
 				statusChan <- 0
 				return
@@ -168,17 +152,14 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 				flusher.Flush()
 			}
 
-			// Reset the timer on each successful write and backoff
-			if !timer.Stop() {
-				select {
-				case <-timer.C: // drain the channel to avoid blocking
-				default:
+			// check if never errored or last error was at least a second ago
+			if lastErr.Equal(timeStarted) || time.Now().Sub(lastErr).Seconds() >= 1 {
+				// Reset timer on successful read/write
+				if !timer.Stop() {
+					<-timer.C // drain channel if necessary
 				}
+				timer.Reset(timeoutDuration)
 			}
-			timer.Reset(timeoutDuration)
-
-			// Reset the backoff duration after successful read/write
-			currentBackoff = initialBackoff
 		}
 	}
 }

--- a/proxy/proxy_stream.go
+++ b/proxy/proxy_stream.go
@@ -14,6 +14,8 @@ import (
 )
 
 func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, resp *http.Response, r *http.Request, w http.ResponseWriter, statusChan chan int) {
+	defer close(statusChan)
+
 	bufferMbInt, err := strconv.Atoi(os.Getenv("BUFFER_MB"))
 	if err != nil || bufferMbInt < 0 {
 		bufferMbInt = 0

--- a/proxy/proxy_stream.go
+++ b/proxy/proxy_stream.go
@@ -98,6 +98,30 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 
 	returnStatus := 0
 
+	// Backoff settings
+	initialBackoff := 200 * time.Millisecond
+	maxBackoff := time.Duration(timeoutSecond-1) * time.Second
+	currentBackoff := initialBackoff
+
+	contextSleep := func() bool {
+		select {
+		case <-time.After(currentBackoff):
+			currentBackoff *= 2
+			if currentBackoff > maxBackoff {
+				currentBackoff = maxBackoff
+			}
+			return true
+		case <-ctx.Done():
+			utils.SafeLogf("Context canceled for stream: %s\n", r.RemoteAddr)
+			statusChan <- 0
+			return false
+		case <-timer.C:
+			utils.SafeLogf("Timeout reached while trying to stream: %s\n", r.RemoteAddr)
+			statusChan <- returnStatus
+			return false
+		}
+	}
+
 	for {
 		select {
 		case <-ctx.Done():
@@ -123,7 +147,9 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 				returnStatus = 2
 
 				utils.SafeLogf("Retrying same stream until timeout (%d seconds) is reached...\n", timeoutSecond)
-				time.Sleep(time.Millisecond * 500)
+				if !contextSleep() {
+					return
+				}
 
 				continue
 			case err != nil:
@@ -136,7 +162,9 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 				}
 
 				utils.SafeLogf("Retrying same stream until timeout (%d seconds) is reached...\n", timeoutSecond)
-				time.Sleep(time.Millisecond * 500)
+				if !contextSleep() {
+					return
+				}
 
 				continue
 			}
@@ -153,13 +181,18 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 			}
 
 			// check if never errored or last error was at least a second ago
-			if lastErr.Equal(timeStarted) || time.Now().Sub(lastErr).Seconds() >= 1 {
+			if lastErr.Equal(timeStarted) || time.Since(lastErr) >= time.Second {
 				// Reset timer on successful read/write
 				if !timer.Stop() {
-					<-timer.C // drain channel if necessary
+					select {
+					case <-timer.C:
+					default:
+					}
 				}
 				timer.Reset(timeoutDuration)
 			}
+
+			currentBackoff = initialBackoff
 		}
 	}
 }

--- a/proxy/proxy_stream.go
+++ b/proxy/proxy_stream.go
@@ -107,17 +107,32 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 
 	contextSleep := func(ctx context.Context, timer *time.Timer) {
 		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
 		case <-time.After(currentBackoff):
 			currentBackoff *= 2
 			if currentBackoff > maxBackoff {
 				currentBackoff = maxBackoff
 			}
-		case <-ctx.Done():
-		case <-timer.C:
 		}
 	}
 
+	readChan := make(chan struct {
+		n   int
+		err error
+	}, 1)
+	defer close(readChan)
+
 	for {
+		go func() {
+			n, err := resp.Body.Read(buffer)
+			readChan <- struct {
+				n   int
+				err error
+			}{n, err}
+		}()
+
 		select {
 		case <-ctx.Done():
 			utils.SafeLogf("Context canceled for stream: %s\n", r.RemoteAddr)
@@ -127,10 +142,9 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 			utils.SafeLogf("Timeout reached while trying to stream: %s\n", r.RemoteAddr)
 			statusChan <- returnStatus
 			return
-		default:
-			n, err := resp.Body.Read(buffer)
+		case result := <-readChan:
 			switch {
-			case err == io.EOF:
+			case result.err == io.EOF:
 				lastErr = time.Now()
 				if utils.EOFIsExpected(resp) || timeoutSecond == 0 {
 					utils.SafeLogf("Stream ended (expected EOF reached): %s\n", r.RemoteAddr)
@@ -145,7 +159,7 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 				contextSleep(ctx, timer)
 
 				continue
-			case err != nil:
+			case result.err != nil:
 				lastErr = time.Now()
 				utils.SafeLogf("Error reading stream: %s\n", err.Error())
 				returnStatus = 1
@@ -160,7 +174,7 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 				continue
 			}
 
-			if _, err := w.Write(buffer[:n]); err != nil {
+			if _, err := w.Write(buffer[:result.n]); err != nil {
 				utils.SafeLogf("Error writing to response: %s\n", err.Error())
 				statusChan <- 0
 				return

--- a/proxy/proxy_stream.go
+++ b/proxy/proxy_stream.go
@@ -140,7 +140,6 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 		select {
 		case <-ctx.Done():
 			utils.SafeLogf("Context canceled for stream: %s\n", r.RemoteAddr)
-			statusChan <- 0
 			return
 		case <-timer.C:
 			utils.SafeLogf("Timeout reached while trying to stream: %s\n", r.RemoteAddr)

--- a/proxy/proxy_stream.go
+++ b/proxy/proxy_stream.go
@@ -73,10 +73,13 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 		return
 	}
 
-	instance.Cm.UpdateConcurrency(m3uIndex, true)
-	defer instance.Cm.UpdateConcurrency(m3uIndex, false)
+	cm := instance.Cm
+
+	cm.UpdateConcurrency(m3uIndex, true)
 
 	defer func() {
+		cm.UpdateConcurrency(m3uIndex, false)
+
 		buffer = nil
 		if flusher, ok := w.(http.Flusher); ok {
 			flusher.Flush()

--- a/proxy/proxy_stream.go
+++ b/proxy/proxy_stream.go
@@ -170,7 +170,6 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 			}
 
 			if _, err := w.Write(buffer[:n]); err != nil {
-				lastErr = time.Now()
 				utils.SafeLogf("Error writing to response: %s\n", err.Error())
 				statusChan <- 0
 				return

--- a/proxy/proxy_stream.go
+++ b/proxy/proxy_stream.go
@@ -103,22 +103,15 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 	maxBackoff := time.Duration(timeoutSecond-1) * time.Second
 	currentBackoff := initialBackoff
 
-	contextSleep := func() bool {
+	contextSleep := func(ctx context.Context, timer *time.Timer) {
 		select {
 		case <-time.After(currentBackoff):
 			currentBackoff *= 2
 			if currentBackoff > maxBackoff {
 				currentBackoff = maxBackoff
 			}
-			return true
 		case <-ctx.Done():
-			utils.SafeLogf("Context canceled for stream: %s\n", r.RemoteAddr)
-			statusChan <- 0
-			return false
 		case <-timer.C:
-			utils.SafeLogf("Timeout reached while trying to stream: %s\n", r.RemoteAddr)
-			statusChan <- returnStatus
-			return false
 		}
 	}
 
@@ -147,9 +140,7 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 				returnStatus = 2
 
 				utils.SafeLogf("Retrying same stream until timeout (%d seconds) is reached...\n", timeoutSecond)
-				if !contextSleep() {
-					return
-				}
+				contextSleep(ctx, timer)
 
 				continue
 			case err != nil:
@@ -162,9 +153,7 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 				}
 
 				utils.SafeLogf("Retrying same stream until timeout (%d seconds) is reached...\n", timeoutSecond)
-				if !contextSleep() {
-					return
-				}
+				contextSleep(ctx, timer)
 
 				continue
 			}
@@ -189,9 +178,9 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 					}
 				}
 				timer.Reset(timeoutDuration)
-			}
 
-			currentBackoff = initialBackoff
+				currentBackoff = initialBackoff
+			}
 		}
 	}
 }

--- a/proxy/proxy_stream.go
+++ b/proxy/proxy_stream.go
@@ -14,7 +14,7 @@ import (
 )
 
 func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, resp *http.Response, r *http.Request, w http.ResponseWriter, statusChan chan int) {
-	defer close(statusChan)
+	debug := os.Getenv("DEBUG") == "true"
 
 	bufferMbInt, err := strconv.Atoi(os.Getenv("BUFFER_MB"))
 	if err != nil || bufferMbInt < 0 {
@@ -73,13 +73,15 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 		return
 	}
 
-	cm := instance.Cm
-
-	cm.UpdateConcurrency(m3uIndex, true)
+	instance.Cm.UpdateConcurrency(m3uIndex, true)
+	defer func() {
+		if debug {
+			utils.SafeLogf("[DEBUG] Defer executed for stream: %s\n", r.RemoteAddr)
+		}
+		instance.Cm.UpdateConcurrency(m3uIndex, false)
+	}()
 
 	defer func() {
-		cm.UpdateConcurrency(m3uIndex, false)
-
 		buffer = nil
 		if flusher, ok := w.(http.Flusher); ok {
 			flusher.Flush()
@@ -125,7 +127,6 @@ func (instance *StreamInstance) ProxyStream(ctx context.Context, m3uIndex int, r
 		n   int
 		err error
 	}, 1)
-	defer close(readChan)
 
 	for {
 		go func() {

--- a/store/concurrency.go
+++ b/store/concurrency.go
@@ -9,7 +9,7 @@ import (
 )
 
 type ConcurrencyManager struct {
-	mu    sync.RWMutex
+	mu    sync.Mutex
 	count map[int]int
 }
 
@@ -32,8 +32,9 @@ func (cm *ConcurrencyManager) Decrement(m3uIndex int) {
 }
 
 func (cm *ConcurrencyManager) GetCount(m3uIndex int) int {
-	cm.mu.RLock()
-	defer cm.mu.RUnlock()
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+
 	return cm.count[m3uIndex]
 }
 

